### PR TITLE
Feature/add proto attrs

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

**Comment by reyoung** [comment]
CI error is because of protobuf's issue https://github.com/google/protobuf/issues/2425

File: cmake/flags.cmake [hunk 1]
```diff
@@ -101,15 +101,16 @@ set(COMMON_FLAGS
     -fPIC
     -fno-omit-frame-pointer
     -Wall
-    -Wextra
```
Comment (wangkuiyi) [inline cmake/flags.cmake]: Why remove this check?

File: cmake/generic.cmake [hunk 2]
```diff
@@ -331,3 +331,42 @@ function(go_test TARGET_NAME)
   add_custom_target(${TARGET_NAME} ALL DEPENDS ${TARGET_NAME}_timestamp ${go_test_DEPS})
   add_test(${TARGET_NAME} ${CMAKE_CURRENT_BINARY_DIR}/${TARGET_NAME})
 endfunction(go_test)
+
+# go_extern will download extern go project.
+# go_extern(target_name extern_source)
+# go_extern(go_redis github.com/hoisie/redis)
+function(go_extern TARGET_NAME)
```
Comment (wangkuiyi) [inline cmake/generic.cmake]: Why we need this function in generic.cmake?

File: cmake/generic.cmake [hunk 3]
```diff
@@ -331,3 +331,42 @@ function(go_test TARGET_NAME)
...
+  add_custom_target(${TARGET_NAME} env GOPATH=${GOPATH} ${CMAKE_Go_COMPILER} get ${ARGN})
+endfunction(go_extern)
+
+
+function(generate_protobuf_cpp SRCS HDRS)
```
Comment (wangkuiyi) [inline cmake/generic.cmake]: According to the documented convention at the beginning of this header file, what we need is `proto_library`.  But the adding of proto_libary seems should be in a separate PR that fixes https://github.com/PaddlePaddle/Paddle/issues/2567.

File: paddle/framework/error.h [hunk 4]
```diff
@@ -14,19 +14,18 @@ limitations under the License. */
 
 #pragma once
 
-#include <glog/logging.h>
```
Comment (wangkuiyi) [inline paddle/framework/error.h]: Let's try don't use the string typed error until we know we must use it.  I had this idea because at least in @jacquesqiao 's PR https://github.com/PaddlePaddle/Paddle/pull/2629#discussion_r124430845, we don't need this error type.

File: paddle/framework/attr_test.proto [hunk 5]
```diff
@@ -0,0 +1,8 @@
+syntax="proto3";
```
Comment (wangkuiyi) [inline paddle/framework/attr_test.proto]: I understand unit test source files are named something like `xxx_test.cc`, but what is `xxx_test.proto` for?

File: paddle/framework/attr_helper.h [hunk 6]
```diff
@@ -0,0 +1,192 @@
+#pragma once
```
Comment (wangkuiyi) [inline paddle/framework/attr_helper.h]: I don't quite understand what goal all the code here is trying to achieve. It might be easier if there is a design doc like [paddle/framework/variable.md](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.md) for [paddle/framework/variable.h](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.h).

**Comment by reyoung** [comment]
Closed, AttributeReader move to PR #2604
